File: Unclassified/173BinarySearchTreeIterator.cpp
#include <iostream>
#include <vector>
#include <stack>
#include <queue>
#include <algorithm>
#include <string>
#include <unordered_set>
#include <unordered_map>

// #include "nestedVectorIntVisualization.h"
// #include "treeVisualizationinPreorder.h"

using namespace std;


// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// next() and hasNext() should run in average O(1) time and uses O(h) memory
// where h is the height of the tree.
class BSTIterator {
public:
    BSTIterator(TreeNode* root) {
        TreeNode* tempNode = root;
        while (tempNode != nullptr)
        {
            nodeStack.push(tempNode);
            tempNode = tempNode->left;
        }
    }

    /** @return the next smallest number */
    int next() {
        TreeNode* tempNode = nodeStack.top();
        nodeStack.pop();
        int minimum = tempNode->val;
        // Don't forget to push the right branch into stack
        if (tempNode->right != nullptr)
        {
            tempNode = tempNode->right;
            while (tempNode != nullptr)
            {
                nodeStack.push(tempNode);
                tempNode = tempNode->left;
            }
        }
        return minimum;
    }

    /** @return whether we have a next smallest number */
    bool hasNext() {
        if (nodeStack.empty())
            return false;
        else
            return true;
    }
private:
    stack<TreeNode*> nodeStack;
};

/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator* obj = new BSTIterator(root);
 * int param_1 = obj->next();
 * bool param_2 = obj->hasNext();
 */

int main()
{
    TreeNode* root = new TreeNode(7);
    TreeNode* t11 = new TreeNode(3);
    TreeNode* t12 = new TreeNode(15);
    TreeNode* t21 = new TreeNode(9);
    TreeNode* t22 = new TreeNode(20);

    root->left = t11;
    t12->left = t21;
    t12->right = t22;

    BSTIterator* obj = new BSTIterator(root);
    int res1 = obj->next();

    cout << res1 << endl;

    if (obj->hasNext())
        cout << "has next small value\n";
    else
        cout << "NO\n";
}

File: Python Solutions/173. Binary Search Tree Iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class BSTIterator:

    def __init__(self, root: Optional[TreeNode]):
        """
        We need a stack to simulate the in-order traversal.
        The stack just keep tracking of the nodes along the left branch all the way to leaf.
        """
        self.stack = []
        while root:
            self.stack.append(root)
            root = root.left

    def next(self) -> int:
        """
        The top element of the stack is the next element. 
        However, every time we pop out the top element, we need to think about the case where the current node (top element) has right branch.
        The right branch's leftmost leaf noode is the next element, which means we should push this to stack for the future usage.
        """
        nextNode = self.stack.pop()
        value = nextNode.val
        if nextNode.right:
            nextNode = nextNode.right
            while nextNode:
                self.stack.append(nextNode)
                nextNode = nextNode.left
        return value

    def hasNext(self) -> bool:
        """
        If the stack is empty --> no next element. Otherwise, has next one.
        """
        return len(self.stack) != 0


# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()